examples/ipfs-private.rs

```diff
@@ -0,0 +1,354 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! A minimal node that can interact with ipfs
+//!
+//! This node implements the gossipsub, ping and identify protocols. It supports
+//! the ipfs private swarms feature by reading the pre shared key file `swarm.key`
+//! from the IPFS_PATH environment variable or from the default location.
+//!
+//! You can pass any number of nodes to be dialed.
+//!
+//! On startup, this example will show a list of addresses that you can dial
+//! from a go-ipfs or js-ipfs node.
+//!
+//! You can ping this node, or use pubsub (gossipsub) on the topic "chat". For this
+//! to work, the ipfs node needs to be configured to use gossipsub.
+use async_std::{io, task};
+use futures::{future, prelude::*};
+use libp2p::core::transport::upgrade::Version;
+use libp2p::identify::{Identify, IdentifyEvent};
+use libp2p::pnet::PnetConfig;
+use libp2p::secio::SecioConfig;
+use libp2p::tcp::TcpConfig;
+use libp2p::yamux::Config as YamuxConfig;
+use libp2p::Transport;
+use libp2p::TransportError;
+use libp2p::{
+    gossipsub::{self, Gossipsub, GossipsubConfigBuilder, GossipsubEvent},
+    identity,
+    ping::{self, Ping, PingConfig, PingEvent},
+    pnet::PreSharedKey,
+    swarm::NetworkBehaviourEventProcess,
+    Multiaddr, NetworkBehaviour, PeerId, Swarm,
+};
+use libp2p_core::either::{EitherError, EitherFuture, EitherListenStream, EitherOutput};
+use libp2p_core::StreamMuxer;
+use multiaddr::Protocol;
+use std::str::FromStr;
+use std::time::Duration;
+use std::{env, fs, path::Path};
+use std::{
+    error::Error,
+    task::{Context, Poll},
+};
+
+/// An Either combination of two transports
+///
+/// this is useful if you have a different transport depending on configuration.
+/// In this example we have a different transport type depending on whether private swarms
+/// are enabled or not.
+#[derive(Debug, Copy, Clone)]
+pub enum EitherTransport<A, B> {
```

Comment:
We should probably get this into `either.rs`. @tomaka, what do you think?

Reply:
Agree

Reply:
Moved it as it is to core/src/either.rs. I think we don't need .left() and .right() methods on the transport trait.